<!DOCTYPE html>

<html lang="en-US" xml:lang="en-US">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>2.CNV calling</title>

<script src="site_libs/header-attrs-2.24/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MAAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-atom"></span>
     
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="1.Peak-calling.html">1.Peak calling</a>
    </li>
    <li>
      <a href="2.CNV-calling.html">2.CNV calling</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">3.SNV calling</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="3.SNV-calling---a.-normal-panel.html">a.Normal panel</a>
        </li>
        <li>
          <a href="3.SNV-calling---HaplotypeCaller.html">b.SNV panel</a>
        </li>
        <li>
          <a href="3.SNV-calling---c.-Tumor-cell-SNV.html">c.Tumor cell SNV</a>
        </li>
        <li>
          <a href="4.Generate-SNV-matrix.html">d.Generate SNV matrix</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="5.SNV-RPCA.html">4.RPCA</a>
    </li>
    <li>
      <a href="6.MAAS.html">5. Running MAAS</a>
    </li>
    <li>
      <a href="7.Clustering.html">6.Clustering</a>
    </li>
    <li>
      <a href="8.Functional-enrichment.html">7.Functional enrichment</a>
    </li>
  </ul>
</li>
<li>
  <a href="Contact.html">
    <span class="fa fa-envelope"></span>
     
    Contact
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">2.CNV calling</h1>

</div>


<p>Here we use the R package <a
href="https://github.com/colomemaria/epiAneufinder">epiAneufinder</a> to
call copy numbers for each cell.</p>
<pre class="r"><code># Clear workspace and load libraries
rm(list = ls())
library(epiAneufinder)</code></pre>
<pre class="r"><code>epiAneufinder(input=&quot;./CRC/CRC-3-11773/raw_data/CRC-3-11773/outs/fragments.tsv.gz&quot;,
              outdir=&quot;./&quot;,
              blacklist=&quot;./hg38-blacklist.v2.bed&quot;,
              windowSize = 1e5,
              genome = &quot;BSgenome.Hsapiens.UCSC.hg38&quot;,
              exclude = c(&#39;chrX&#39;,&#39;chrY&#39;,&#39;chrM&#39;),
              reuse.existing = TRUE,
              title_karyo=&quot;&quot;,
              ncores = 60,
              minFrags = 15000)</code></pre>
<hr />
<p>After finishing running <code>epiAneufinder</code>, you will get the
a folder named <code>epiAneufinder_results</code> in the current
directory. Then we should perform CNV visualization and clustering to
identify aneuploid tumor cells.</p>
<pre class="r"><code>library(data.table)
library(dplyr)
library(ComplexHeatmap)
library(circlize)

#Obtain the range of chromosomes splitted to short (p) and long arms (q)
chrom.range &lt;- fread(&quot;http://hgdownload.cse.ucsc.edu/goldenpath/hg38/database/cytoBand.txt.gz&quot;,
           col.names = c(&quot;chrom&quot;,&quot;Start&quot;,&quot;End&quot;,&quot;arm&quot;,&quot;gieStain&quot;))
chrom.list &lt;- paste0(&quot;chr&quot;,c(1:22, &quot;X&quot;))
chrom.range &lt;- chrom.range[which(chrom.range$chrom %in% chrom.list),]
chrom.range &lt;- dplyr::select(chrom.range, -c(gieStain))
chrom.range$arm &lt;- substr(chrom.range$arm, 1,1)
chrom.range &lt;- chrom.range %&gt;%
  mutate(chrom_band = paste(chrom, arm, sep = &quot;&quot;))
chrom.range &lt;- chrom.range %&gt;%
  group_by(chrom_band) %&gt;%
  dplyr::summarise(start = min(Start),
            end = max(End))
chrom.range$chrom &lt;- gsub(&quot;([[:alnum:]]+)[[:alpha:]]+$&quot;, &quot;\\1&quot;, chrom.range$chrom_band)</code></pre>
<p>Then we manipulate the epiAneufinder results to obatin a cell-by-CNV
matrix.</p>
<pre class="r"><code>#  One can run the following code directly.
df &lt;- read.table(&quot;./epiAneufinder_results/results_table.tsv&quot;, row.names = 1)
input.range &lt;- df[,1:3]
df$range &lt;- paste0(df$seq, &quot;:&quot;, df$start, &quot;-&quot;, df$end)
df &lt;- dplyr::select(df, -c(seq, start, end))
rownames(df) &lt;- df$range
df &lt;- dplyr::select(df, -c(range))
df &lt;- as.data.frame(t(df))
deter.arm &lt;- c()
for(i in 1:nrow(input.range)){
  chr.tmp &lt;- input.range$seq[i]
  start.tmp &lt;- input.range$start[i]
  end.tmp &lt;- input.range$end[i]
  chrom.range.tmp &lt;- chrom.range[which(chrom.range$chrom == chr.tmp &amp;
                                         chrom.range$start &lt;= start.tmp &amp;
                                         chrom.range$end &gt;= end.tmp),]
  deter.arm[i] &lt;- chrom.range.tmp$chrom_band
}
deter.arm &lt;- substr(deter.arm, nchar(deter.arm), nchar(deter.arm))
colnames(df) &lt;- paste0(colnames(df), &quot;(&quot;, deter.arm, &quot;)&quot;)
columns &lt;- colnames(df)</code></pre>
<p>To identify tumor cells, we perform K-means clustering to distinguish
aneuploidy from normal cells that have low frequency of CNVs and
low-degree of clonal copy numbers.</p>
<pre class="r"><code># K-means clustering
set.seed(2)
clu &lt;- kmeans(df, centers = 3)$cluster
clu &lt;- as.data.frame(clu)
clu &lt;- clu[order(clu$clu),,drop = F]
df &lt;- df[rownames(clu),]

# Visualization
chr_full_name &lt;- gsub(&quot;chr([0-9XYM]+):([0-9]+)-([0-9]+)\\(([pq])\\).*&quot;, &quot;chr\\1\\4&quot;, colnames(df))
ann.col &lt;- c(rep(c(&quot;black&quot;, &quot;#E0E0E0&quot;), ceiling(length(unique(chr_full_name)) / 2)))
names(ann.col) &lt;- unique(chr_full_name)
ann.row &lt;- c(&quot;#D95F02&quot;, &quot;#1B9E77&quot;, &quot;#B07164&quot;)
names(ann.row) &lt;- c(&quot;1&quot;, &quot;2&quot;, &quot;3&quot;)


heat.col &lt;- colorRamp2(c(0, 1, 2), c(&quot;#408CBF&quot;, &quot;white&quot;, &quot;#D15848&quot;))
pdf(&quot;totalcell.cnv.heatmap.pdf&quot;, width = 9, height = 4.5)
ha &lt;- HeatmapAnnotation(chr = chr_full_name,
                        col = list(chr = ann.col),
                        show_annotation_name = F,
                        show_legend = F)
row_ha &lt;- rowAnnotation(Type = clu$clu,
                        col = list(Type = ann.row),
                        show_annotation_name = F,
                        show_legend = F)
Heatmap(as.matrix(df), name=&quot;CNV&quot;,
        top_annotation = ha,
        left_annotation = row_ha,
        col = heat.col,
        column_names_gp = gpar(fontsize=10),
        cluster_rows = F,
        cluster_columns = F,
        show_row_dend = F,
        show_row_names = F,
        show_column_names = F)
dev.off()</code></pre>
<p><img src="http://bioinfo.szbl.ac.cn/share/MAAS_data/totalcell.cnv.heatmap.png" width="75%" height="50%" style="display: block; margin: auto;" />
According to the heatmap, we can see that cluster 1 and cluster 3
exhibit aneuploidy and clonal pattern of copy numbers, while cluster 2
demonstrate low-degree of CNVs. Hence, we annotate cells in cluster 1
and cluster 3 as tumor cells.</p>
<pre class="r"><code># Cross validate cell annotation by ArchR
normal.cell &lt;- read.csv(&quot;../1.Peak_calling/normal.cell.txt&quot;, sep = &#39;\t&#39;, header = T)
normal.cell$V1 &lt;- gsub(&quot;CRC#&quot;, &quot;&quot;, normal.cell$V1)
# Check if aneuploidy cells are annotated as normal cells by ArchR
intersect(rownames(cancer.df), normal.cell$V1)</code></pre>
<pre class="r"><code># Save the cell-by-CNV matrix of tumor cells
cancer.df &lt;- df[rownames(clu[which(clu$clu != 2),,drop = F]),]
write.csv(cancer.df, &quot;cancer.cnv.profile.csv&quot;)</code></pre>
<pre class="r"><code># Save tumor barcodes
x &lt;- rownames(cancer.df)
x &lt;- gsub(&quot;cell.&quot;, &quot;&quot;, x)
x &lt;- gsub(&quot;\\.&quot;, &quot;-&quot;, x)
x &lt;- data.frame(barcode = x, barcode2 = x)
 # This file can be used for bam file split according to barcode names
write.table(x, &quot;cancer.barcode.txt&quot;, sep = &#39;\t&#39;, quote = F, col.names = F, row.names = F)</code></pre>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
