<!DOCTYPE html>

<html lang="en-US" xml:lang="en-US">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>8.Functional enrichment</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MAAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-atom"></span>
     
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="1.Peak-calling.html">1.Peak calling</a>
    </li>
    <li>
      <a href="2.CNV-calling.html">2.CNV calling</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">3.SNV calling</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="3.SNV-calling---a.-normal-panel.html">a.Normal panel</a>
        </li>
        <li>
          <a href="3.SNV-calling---HaplotypeCaller.html">b.SNV panel</a>
        </li>
        <li>
          <a href="3.SNV-calling---c.-Tumor-cell-SNV.html">c.Tumor cell SNV</a>
        </li>
        <li>
          <a href="4.Generate-SNV-matrix.html">d.Generate SNV matrix</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="5.SNV-RPCA.html">4.RPCA</a>
    </li>
    <li>
      <a href="6.MAAS.html">5. Running MAAS</a>
    </li>
    <li>
      <a href="7.Clustering.html">6.Clustering</a>
    </li>
    <li>
      <a href="8.Functional-enrichment.html">7.Functional enrichment</a>
    </li>
  </ul>
</li>
<li>
  <a href="Contact.html">
    <span class="fa fa-envelope"></span>
     
    Contact
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">8.Functional enrichment</h1>

</div>


<p>We can finally analyze the biological function enrichment between the
subpopulations based on the cancer hallmark signatures, which can be
downloaded from <a
href="https://www.gsea-msigdb.org/gsea/msigdb/">MSigDB</a>.</p>
<pre class="r"><code># Clear working space and load libraries
rm(list = ls())
library(limma)
library(GSEABase)
library(GSVA)
library(pheatmap)
clusterFile &lt;- &quot;../6.Cluster/maas.pca.rds&quot;
gmtFile &lt;- &quot;h.all.v2023.1.Hs.symbols.gmt&quot;</code></pre>
<pre class="r"><code># Load gene activity score
data &lt;- readRDS(&quot;../1.Peak_calling/geneScoreMatrix.rds&quot;)
colnames(data) &lt;- gsub(&quot;CRC#&quot;, &quot;&quot;, colnames(data))
cluster &lt;- readRDS(clusterFile)
data &lt;- data[,rownames(cluster)]</code></pre>
<pre class="r"><code># Single cell GSVA
geneSets &lt;- getGmt(gmtFile, geneIdType=SymbolIdentifier())
gsvaResult &lt;- gsva(as.matrix(data), geneSets, min.sz = 10, verbose = T, parallel.sz = 50)
# gsvaOut &lt;- rbind(id = colnames(gsvaResult), gsvaResult)
# write.table(gsvaOut, file=&quot;gsvaOut.txt&quot;, sep=&quot;\t&quot;, quote=F, col.names=F)

gsvaResult &lt;- as.data.frame(t(gsvaResult))
colnames(gsvaResult) &lt;- gsub(&quot;HALLMARK_&quot;, &quot;&quot;, colnames(gsvaResult))
sameSample &lt;- intersect(row.names(gsvaResult), row.names(cluster))
gsvaResult1 &lt;- gsvaResult[sameSample,,drop=F]
cluster &lt;- cluster[sameSample,,drop=F]
gsvaCluster &lt;- cbind(gsvaResult1, Subpopulation = cluster$Subpopulation)</code></pre>
<pre class="r"><code># Differential analysis
adj.P.Val.Filter=0.05
allType=as.vector(gsvaCluster$Subpopulation)
comp=combn(levels(factor(allType)), 2)
for(i in 1:ncol(comp)){
  #样品分组
  treat=gsvaCluster[gsvaCluster$Subpopulation==comp[2,i],]
  con=gsvaCluster[gsvaCluster$Subpopulation==comp[1,i],]
  data=rbind(con, treat)
  #差异分析
  Type=as.vector(data$Subpopulation)
  ann=data.frame(data[,ncol(data)])
  row.names(ann)=row.names(data)
  names(ann)=&quot;Subpopulation&quot;
  data=t(data[,-(ncol(data))])
  design=model.matrix(~0+factor(Type))
  colnames(design)=levels(factor(Type))
  fit=lmFit(data, design)
  contrast=paste0(comp[2,i], &quot;-&quot;, comp[1,i])
  cont.matrix=makeContrasts(contrast, levels=design)
  fit2=contrasts.fit(fit, cont.matrix)
  fit2=eBayes(fit2)
  
  #输出所有通路的差异情况
  allDiff=topTable(fit2,adjust=&#39;fdr&#39;,number=200000)
  allDiffOut=rbind(id=colnames(allDiff),allDiff)
  write.table(allDiffOut, file=paste0(contrast, &quot;.all.txt&quot;), sep=&quot;\t&quot;, quote=F, col.names=F)
  
  #输出显著的差异
  diffSig=allDiff[with(allDiff, (abs(logFC) &gt; 0.075 &amp; adj.P.Val &lt; adj.P.Val.Filter )), ]
  diffSigOut=rbind(id=colnames(diffSig),diffSig)
  write.table(diffSigOut, file=paste0(contrast, &quot;.diff.txt&quot;), sep=&quot;\t&quot;, quote=F, col.names=F)
  #聚类颜色
  my_color &lt;- c(&quot;#5050FF&quot;, &quot;#CE3D32&quot;, &quot;#749B58&quot;, &quot;#F0E685&quot;, &quot;#00BFC4&quot;, &quot;#B79F00&quot;, &quot;#00BA38&quot;)
  heatmap.BlBkRd &lt;- c(&quot;#00DDFF&quot;,&quot;#32ABAA&quot;,&quot;#125456&quot;,&quot;#000000&quot;,&quot;#510000&quot;,&quot;#A20000&quot;,&quot;#F30000&quot;)
  ann_colors=list()
  CluCol=my_color[1:length(levels(factor(allType)))]
  names(CluCol)=levels(factor(allType))
  ann_colors[[&quot;Subpopulation&quot;]]=CluCol[c(comp[1,i], comp[2,i])]

  #绘制差异通路
  termNum=30
  diffTermName=as.vector(rownames(diffSig))
  diffLength=length(diffTermName)
  if(diffLength&lt;termNum){termNum=diffLength}
  hmGene=diffTermName[1:termNum]
  hmExp=data[hmGene,]
  
  p1 &lt;- pheatmap::pheatmap(hmExp,
           annotation=ann,
           annotation_colors = ann_colors,
           color = colorRampPalette(c(&quot;#003A7A&quot;, &quot;white&quot;, &quot;#A20000&quot;))(60),
           cluster_cols =F,
           show_colnames = F,
           # gaps_col=as.vector(cumsum(table(Type))),
           scale=&quot;row&quot;,
           cluster_row = T,
           fontsize = 10,
           fontsize_row=8,
           fontsize_col=10,
           cellheight = 20,
           treeheight_row = 0)
  ggsave(paste0(contrast,&quot;.heatmap_hallmark.pdf&quot;), p1, height = 6, width = 6)
}</code></pre>
<p><img src="http://bioinfo.szbl.ac.cn/share/MAAS_data/S2-S1.heatmap_hallmark.png" width="50%" height="35%" style="display: block; margin: auto;" /></p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
