<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3.SNV calling â€¢ MAAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3.SNV calling">
<meta property="og:description" content="MAAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1.scATAC-seq%20data%20analysis.html">1.scATAC-seq data analysis</a>
    </li>
    <li>
      <a href="../articles/2.CNV%20calling.html">2.CNV calling</a>
    </li>
    <li>
      <a href="../articles/3.SNV%20calling.html">3.SNV calling</a>
    </li>
    <li>
      <a href="../articles/4.Peak%20correction.html">4.Peak correction</a>
    </li>
    <li>
      <a href="../articles/5.SNV%20correction.html">5.SNV correction</a>
    </li>
    <li>
      <a href="../articles/6.MAAS.html">6.MAAS</a>
    </li>
    <li>
      <a href="../articles/7.Clustering.html">7.Clustering</a>
    </li>
    <li>
      <a href="../articles/8.Functional%20enrichment.html">8.Functional enrichment</a>
    </li>
    <li>
      <a href="../articles/Contact.html">Contact</a>
    </li>
    <li>
      <a href="../articles/index.html">Tutorial of MAAS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Larrycpan/MAAS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>3.SNV calling</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Larrycpan/MAAS/blob/HEAD/vignettes/3.SNV%20calling.Rmd" class="external-link"><code>vignettes/3.SNV calling.Rmd</code></a></small>
      <div class="hidden name"><code>3.SNV calling.Rmd</code></div>

    </div>

    
    
<p>We used <a href="https://github.com/cortes-ciriano-lab/SComatic" class="external-link">SComatic</a> to
obtain SNV profile of tumor cells from scATAC-seq data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Create a directory for SNV calling in the terminal</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="fl">3.</span>SNV_calling</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># Step1: split bam files for different cell types</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>utput_dir1<span class="ot">=</span>step1_BamCellTypes</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="sc">$</span>output_dir1</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>SplitBam<span class="sc">/</span>SplitBamCellTypes.py \</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>        <span class="sc">--</span>bam <span class="sc">/</span>Path<span class="sc">/</span>to<span class="sc">/</span>bam<span class="sc">/</span>file \</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>        <span class="sc">--</span>meta ..<span class="sc">/</span><span class="fl">1.</span>Cellanno<span class="sc">/</span>HCC.celltype.txt\</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>        <span class="sc">--</span>id example \</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>        <span class="sc">--</span>min_MQ <span class="dv">30</span> \</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>        <span class="sc">--</span>outdir <span class="sc">$</span>output_dir1</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="do">################################################################################</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co"># Step2: summarize base count</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>REF<span class="ot">=</span><span class="er">/</span>Path<span class="sc">/</span>to<span class="sc">/</span>refdata<span class="sc">-</span>cellranger<span class="sc">-</span>arc<span class="sc">-</span>GRCh38<span class="dv">-2020</span><span class="sc">-</span>A<span class="dv">-2</span>.<span class="fl">0.0</span><span class="sc">/</span>fasta<span class="sc">/</span>genome.fa</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>  </span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>output_dir2<span class="ot">=</span>step2_BaseCellCounts</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="sc">$</span>output_dir2</span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="cf">for</span> bam <span class="cf">in</span> <span class="sc">$</span>(ls <span class="sc">-</span>d .<span class="sc">/</span>step1_BamCellTypes<span class="sc">/</span><span class="er">*</span>bam);do</span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a>        <span class="co"># Cell type</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a>        cell_type<span class="ot">=</span><span class="er">$</span>(basename <span class="sc">$</span>bam <span class="sc">|</span> awk <span class="sc">-</span>F<span class="st">'.'</span> <span class="st">'{print $(NF-1)}'</span>)</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a>        <span class="co"># Temp folder</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>        temp<span class="ot">=</span><span class="er">$</span>output_dir2<span class="sc">/</span>temp_<span class="sc">$</span>{cell_type}</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a>        mkdir <span class="sc">-</span>p <span class="sc">$</span>temp</span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>        <span class="co"># Command line to submit to cluster</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a>        python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>BaseCellCounter<span class="sc">/</span>BaseCellCounter.py <span class="sc">--</span>bam <span class="sc">$</span>bam \</span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>                <span class="sc">--</span>ref <span class="sc">$</span>REF \</span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a>                <span class="sc">--</span>chrom all \</span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a>                <span class="sc">--</span>out_folder <span class="sc">$</span>output_dir2 \</span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a>                <span class="sc">--</span>bed <span class="sc">/</span>Path<span class="sc">/</span>to<span class="sc">/</span>ATAC<span class="sc">/</span>bed<span class="sc">/</span>file \</span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a>                <span class="sc">--</span>min_bq <span class="dv">30</span> \</span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a>                <span class="sc">--</span>min_mq <span class="dv">30</span> \</span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a>                <span class="sc">--</span>tmp_dir <span class="sc">$</span>temp \</span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a>                <span class="sc">--</span>nprocs <span class="dv">30</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a>        rm <span class="sc">-</span>rf <span class="sc">$</span>temp</span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a>done</span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="do">################################################################################</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="co"># Step3: merge counts</span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a>output_dir3<span class="ot">=</span>step3_BaseCellCountsMerged</span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="sc">$</span>output_dir3</span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a></span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a>python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>MergeCounts<span class="sc">/</span>MergeBaseCellCounts.py <span class="sc">--</span>tsv_folder .<span class="sc">/</span>step2_BaseCellCounts <span class="sc">--</span>outfile <span class="sc">$</span>{output_dir3}<span class="sc">/</span>BaseCellCounts.AllCellTypes.tsv</span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="do">################################################################################</span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a></span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="co"># Step4: Mutation detection</span></span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a><span class="co"># Step 4.1</span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a>output_dir4<span class="ot">=</span><span class="er">/</span>step4_VariantCalling</span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="sc">$</span>output_dir4</span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a>sample<span class="ot">=</span>example</span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a>python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>BaseCellCalling<span class="sc">/</span>BaseCellCalling.step1.py \</span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a>                  <span class="sc">--</span>infile step3_BaseCellCountsMerged<span class="sc">/</span>BaseCellCounts.AllCellTypes.tsv \</span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a>                  <span class="sc">--</span>outfile <span class="sc">$</span>{output_dir4}<span class="sc">/</span><span class="er">$</span>{sample} \</span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a>                  <span class="sc">--</span>ref <span class="sc">$</span>REF</span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a></span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a><span class="co"># Step 4.2</span></span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a>PON<span class="ot">=</span><span class="er">/</span>Path<span class="sc">/</span>to<span class="sc">/</span>PoN.scATACseq.hg38.tsv</span>
<span id="cb1-69"><a href="#cb1-69" tabindex="-1"></a></span>
<span id="cb1-70"><a href="#cb1-70" tabindex="-1"></a>python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>BaseCellCalling<span class="sc">/</span>BaseCellCalling.step2.py \</span>
<span id="cb1-71"><a href="#cb1-71" tabindex="-1"></a>                  <span class="sc">--</span>infile <span class="sc">$</span>{output_dir4}<span class="sc">/</span><span class="er">$</span>{sample}.calling.step1.tsv \</span>
<span id="cb1-72"><a href="#cb1-72" tabindex="-1"></a>                  <span class="sc">--</span>outfile <span class="sc">$</span>{output_dir4}<span class="sc">/</span><span class="er">$</span>{sample} \</span>
<span id="cb1-73"><a href="#cb1-73" tabindex="-1"></a>                  <span class="sc">--</span>pon <span class="sc">$</span>PON</span>
<span id="cb1-74"><a href="#cb1-74" tabindex="-1"></a></span>
<span id="cb1-75"><a href="#cb1-75" tabindex="-1"></a><span class="do">################################################################################</span></span>
<span id="cb1-76"><a href="#cb1-76" tabindex="-1"></a></span>
<span id="cb1-77"><a href="#cb1-77" tabindex="-1"></a><span class="co"># Step5: site calling</span></span>
<span id="cb1-78"><a href="#cb1-78" tabindex="-1"></a>STEP4_1<span class="ot">=</span><span class="er">/</span>step4_VariantCalling<span class="sc">/</span><span class="er">$</span>{sample}.calling.step1.tsv</span>
<span id="cb1-79"><a href="#cb1-79" tabindex="-1"></a></span>
<span id="cb1-80"><a href="#cb1-80" tabindex="-1"></a>output_dir5<span class="ot">=</span>step5_UniqueCellCallableSites</span>
<span id="cb1-81"><a href="#cb1-81" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="sc">$</span>output_dir5</span>
<span id="cb1-82"><a href="#cb1-82" tabindex="-1"></a></span>
<span id="cb1-83"><a href="#cb1-83" tabindex="-1"></a><span class="cf">for</span> bam <span class="cf">in</span> <span class="sc">$</span>(ls <span class="sc">-</span>d .<span class="sc">/</span>step1_BamCellTypes<span class="sc">/</span><span class="er">*</span>bam);do</span>
<span id="cb1-84"><a href="#cb1-84" tabindex="-1"></a>            cell_type<span class="ot">=</span><span class="er">$</span>(basename <span class="sc">$</span>bam <span class="sc">|</span> awk <span class="sc">-</span>F<span class="st">'.'</span> <span class="st">'{print $(NF-1)}'</span>)</span>
<span id="cb1-85"><a href="#cb1-85" tabindex="-1"></a>            echo <span class="sc">$</span>cell_type</span>
<span id="cb1-86"><a href="#cb1-86" tabindex="-1"></a></span>
<span id="cb1-87"><a href="#cb1-87" tabindex="-1"></a>            temp<span class="ot">=</span><span class="er">$</span>output_dir5<span class="sc">/</span>temp_<span class="sc">$</span>{cell_type}</span>
<span id="cb1-88"><a href="#cb1-88" tabindex="-1"></a>            mkdir <span class="sc">-</span>p <span class="sc">$</span>temp</span>
<span id="cb1-89"><a href="#cb1-89" tabindex="-1"></a></span>
<span id="cb1-90"><a href="#cb1-90" tabindex="-1"></a>            python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>SitesPerCell<span class="sc">/</span>SitesPerCell.py <span class="sc">--</span>bam <span class="sc">$</span>bam    \</span>
<span id="cb1-91"><a href="#cb1-91" tabindex="-1"></a>                    <span class="sc">--</span>infile .<span class="sc">/</span>step4_VariantCalling<span class="sc">/</span><span class="er">$</span>{sample}.calling.step1.tsv   \</span>
<span id="cb1-92"><a href="#cb1-92" tabindex="-1"></a>                    <span class="sc">--</span>min_mq <span class="dv">30</span> \</span>
<span id="cb1-93"><a href="#cb1-93" tabindex="-1"></a>                    <span class="sc">--</span>ref <span class="sc">$</span>REF \</span>
<span id="cb1-94"><a href="#cb1-94" tabindex="-1"></a>                    <span class="sc">--</span>out_folder <span class="sc">$</span>output_dir5 \</span>
<span id="cb1-95"><a href="#cb1-95" tabindex="-1"></a>                    <span class="sc">--</span>tmp_dir <span class="sc">$</span>temp \</span>
<span id="cb1-96"><a href="#cb1-96" tabindex="-1"></a>                    <span class="sc">--</span>nprocs <span class="dv">60</span></span>
<span id="cb1-97"><a href="#cb1-97" tabindex="-1"></a>            echo</span>
<span id="cb1-98"><a href="#cb1-98" tabindex="-1"></a>done</span>
<span id="cb1-99"><a href="#cb1-99" tabindex="-1"></a></span>
<span id="cb1-100"><a href="#cb1-100" tabindex="-1"></a><span class="do">################################################################################</span></span>
<span id="cb1-101"><a href="#cb1-101" tabindex="-1"></a></span>
<span id="cb1-102"><a href="#cb1-102" tabindex="-1"></a><span class="co"># Step6: get genotypes</span></span>
<span id="cb1-103"><a href="#cb1-103" tabindex="-1"></a></span>
<span id="cb1-104"><a href="#cb1-104" tabindex="-1"></a>output_dir6<span class="ot">=</span>step6_SingleCellAlleles</span>
<span id="cb1-105"><a href="#cb1-105" tabindex="-1"></a>mkdir <span class="sc">-</span>p <span class="sc">$</span>output_dir6</span>
<span id="cb1-106"><a href="#cb1-106" tabindex="-1"></a></span>
<span id="cb1-107"><a href="#cb1-107" tabindex="-1"></a><span class="cf">for</span> bam <span class="cf">in</span> <span class="sc">$</span>(ls <span class="sc">-</span>d .<span class="sc">/</span>step1_BamCellTypes<span class="sc">/</span><span class="er">*</span>bam);do</span>
<span id="cb1-108"><a href="#cb1-108" tabindex="-1"></a></span>
<span id="cb1-109"><a href="#cb1-109" tabindex="-1"></a>        cell_type<span class="ot">=</span><span class="er">$</span>(basename <span class="sc">$</span>bam <span class="sc">|</span> awk <span class="sc">-</span>F<span class="st">'.'</span> <span class="st">'{print $(NF-1)}'</span>)</span>
<span id="cb1-110"><a href="#cb1-110" tabindex="-1"></a></span>
<span id="cb1-111"><a href="#cb1-111" tabindex="-1"></a>        temp<span class="ot">=</span><span class="er">$</span>output_dir6<span class="sc">/</span>temp_<span class="sc">$</span>{cell_type}</span>
<span id="cb1-112"><a href="#cb1-112" tabindex="-1"></a>        mkdir <span class="sc">-</span>p <span class="sc">$</span>temp</span>
<span id="cb1-113"><a href="#cb1-113" tabindex="-1"></a></span>
<span id="cb1-114"><a href="#cb1-114" tabindex="-1"></a>        python <span class="sc">~</span><span class="er">/</span>App<span class="sc">/</span>SComatic<span class="sc">-</span>main<span class="sc">/</span>scripts<span class="sc">/</span>SingleCellGenotype<span class="sc">/</span>SingleCellGenotype.py <span class="sc">--</span>bam <span class="sc">$</span>bam  \</span>
<span id="cb1-115"><a href="#cb1-115" tabindex="-1"></a>                <span class="sc">--</span>infile <span class="sc">$</span>{STEP4_2_pass}   \</span>
<span id="cb1-116"><a href="#cb1-116" tabindex="-1"></a>                <span class="sc">--</span>nprocs <span class="dv">30</span>   \</span>
<span id="cb1-117"><a href="#cb1-117" tabindex="-1"></a>                <span class="sc">--</span>meta <span class="sc">$</span>META   \</span>
<span id="cb1-118"><a href="#cb1-118" tabindex="-1"></a>                <span class="sc">--</span>outfile <span class="sc">$</span>{output_dir6}<span class="sc">/</span><span class="er">$</span>{cell_type}.single_cell_genotype.tsv  \</span>
<span id="cb1-119"><a href="#cb1-119" tabindex="-1"></a>                <span class="sc">--</span>tmp_dir <span class="sc">$</span>temp  \</span>
<span id="cb1-120"><a href="#cb1-120" tabindex="-1"></a>                <span class="sc">--</span>min_mq <span class="dv">30</span> \</span>
<span id="cb1-121"><a href="#cb1-121" tabindex="-1"></a>                <span class="sc">--</span>ref <span class="sc">$</span>REF</span>
<span id="cb1-122"><a href="#cb1-122" tabindex="-1"></a></span>
<span id="cb1-123"><a href="#cb1-123" tabindex="-1"></a>        rm <span class="sc">-</span>rf <span class="sc">$</span>temp</span>
<span id="cb1-124"><a href="#cb1-124" tabindex="-1"></a>done</span></code></pre></div>
<p>After running the pipeling of SComatic, we can generate a cell-by-SNV
matrix for MAAS integration as follows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">genotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"step6_SingleCellAlleles/Tumor.single_cell_genotype.tsv"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">genotype</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"chr"</span></span>
<span><span class="va">genotype1</span> <span class="op">&lt;-</span> <span class="va">genotype</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Num_reads</span> <span class="op">&gt;</span> <span class="fl">2</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^chr"</span>, <span class="va">chr</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">genotype</span><span class="op">$</span><span class="va">ALT_expected</span> <span class="op">!=</span> <span class="st">"."</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genotype1</span> <span class="op">&lt;-</span> <span class="va">genotype1</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>genotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Base_observed</span> <span class="op">==</span> <span class="fu">str_extract</span><span class="op">(</span><span class="va">ALT_expected</span>, <span class="st">"^[^,]+"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">Cell_type_observed</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">Cell_type_expected</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genotype1</span><span class="op">$</span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">genotype1</span><span class="op">$</span><span class="va">chr</span>, <span class="st">":"</span>, <span class="va">genotype1</span><span class="op">$</span><span class="va">Start</span>, <span class="st">":"</span>, <span class="va">genotype1</span><span class="op">$</span><span class="va">REF</span>, <span class="st">"&gt;"</span>, <span class="fu">str_extract</span><span class="op">(</span><span class="va">genotype1</span><span class="op">$</span><span class="va">ALT_expected</span>, <span class="st">"^[^,]+"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/dcast.data.table.html" class="external-link">dcast</a></span><span class="op">(</span><span class="va">genotype1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CB"</span>, <span class="st">"pos"</span>, <span class="st">"genotype"</span><span class="op">)</span><span class="op">]</span>, <span class="va">pos</span> <span class="op">~</span> <span class="va">CB</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">snv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tibble/man/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="va">snv</span>, var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">snv</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">snv1</span> <span class="op">&lt;-</span> <span class="va">snv</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">snv</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span>,<span class="op">]</span>   <span class="co"># Maintain SNVs in at least 10 tumor cells</span></span>
<span><span class="co"># saveRDS(snv1, "snv.mat.rds")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">snv1</span>, <span class="st">"example.snv.csv"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snv1</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;                    AAACGAATCCATGTTT AAACGAATCCGGCGAT AAACTCGAGGAAGACA</span></span>
<span><span class="co">#&gt; chr1:221742291:T&gt;A                0                0                0</span></span>
<span><span class="co">#&gt; chr1:633990:T&gt;G                   0                0                0</span></span>
<span><span class="co">#&gt; chr1:634045:T&gt;G                   0                0                0</span></span>
<span><span class="co">#&gt; chr1:85276655:G&gt;C                 0                0                0</span></span>
<span><span class="co">#&gt; chr16:34586831:A&gt;G                0                0                0</span></span>
<span><span class="co">#&gt;                    AAACTCGAGTTCGACT AAACTCGCATTCTTTG</span></span>
<span><span class="co">#&gt; chr1:221742291:T&gt;A                0                0</span></span>
<span><span class="co">#&gt; chr1:633990:T&gt;G                   0                0</span></span>
<span><span class="co">#&gt; chr1:634045:T&gt;G                   0                1</span></span>
<span><span class="co">#&gt; chr1:85276655:G&gt;C                 0                0</span></span>
<span><span class="co">#&gt; chr16:34586831:A&gt;G                0                0</span></span></code></pre>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kewei Xiong, Ruofan Ding, Lei Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
